doc/man1/mpq-info.1
.\" Copyright (c) 2003-2008 Maik Broemme <mbroemme@plusserver.de>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.TH mpq-tools 1 2008-02-10 "The MoPaQ archive library"
.SH NAME
mpq-info \- utility to print useful archive information of the given mopaq (mpq) archive.
.SH SYNOPSIS
.B mpq-info
[options] [file...]
.SH DESCRIPTION
.PP
\fImpq-info\fP is a utility to print some detailed archive information like number of files, block size, compression ratio and so on of the given mpq archive to standard output. It is a fast and leightweight application which can be easily integrated in scripts. It accepts multiple files as command line arguments.
.SH OPTIONS
.l
\fImpq-info\fP accepts the following options:
.TP 8
.B  \-h|\-\-help
.ti 15
Print the usage message on the standard output.
.TP 8
.B  \-v|\-\-version
.ti 15
Print the currently installed version on the standard output.
.SH SEE ALSO
\fBmpq-extract\fR(1), \fBlibmpq-config\fR(1)
.SH AUTHOR
Check documentation.
.TP
libmpq is (c) 2003-2008
.B Maik Broemme <mbroemme@plusserver.de>
.PP
The above e-mail address can be used to send bug reports, feedbacks or library enhancements.


NEWS
Changes version 0.4.1 (2008-04-02)
==================================

    * memory, speed and stability improvements.

    * split sourcecode into library and utility package.

Changes version 0.4.0 (2008-03-31)
==================================

    * added robust error handling to make the library and extracting
      utility more stable.

    * added c++ bindings to public include and moved internal defines
      and functions to private include.

    * added support for 64-bit architectures and removed any stupid
      pointer to int arithmetics.

    * added much better member names to the structures to make
      developers able to understand the code on reading.

    * added full api documentation using manual pages.

    * added full extraction support for all blizzard titles until
      warcraft 3 - the frozen throne.

    * added support for single sector and stored files (neither
      compressed nor imploded).

    * added support for files which have compressed size greater than
      uncompressed size.

    * removed the external listfile database support from 0.3.0 it
      was a weird implementation.

Changes version 0.3.0 (2004-02-12)
==================================

    * added listfile database support from external files.

Initial version 0.2.1 (2004-01-17)
==================================

    * first version which was able to extract some of the older mopaq
      archives.


src/mpq-info.c
/*
 *  mpq-info.c -- functions for information about the given mpq archive.
 *
 *  Copyright (c) 2003-2007 Maik Broemme <mbroemme@plusserver.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/* generic includes. */
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* libmpq includes. */
#include "libmpq/mpq.h"

/* zlib includes. */
#include <zlib.h>

/* mpq-tools configuration includes. */
#include "config.h"

/* define new print functions for error. */
#define ERROR(...) fprintf(stderr, __VA_ARGS__);

/* define new print functions for notification. */
#define NOTICE(...) printf(__VA_ARGS__);

/* this function show the usage. */
int mpq_info__usage(char *program_name) {

	/* show the help. */
	NOTICE("Usage: %s [OPTION] [ARCHIVE]...\n", program_name);
	NOTICE("Displays information of a mpq-archive. (Example: %s d2speech.mpq)\n", program_name);
	NOTICE("\n");
	NOTICE("  -h, --help		shows this help screen\n");
	NOTICE("  -v, --version		shows the version information\n");
	NOTICE("\n");
	NOTICE("Please report bugs to the appropriate authors, which can be found in the\n");
	NOTICE("version information. All other things can be send to <%s>\n", PACKAGE_BUGREPORT);

	/* if no error was found, return zero. */
	return 0;
}

/* this function shows the version information. */
int mpq_info__version(char *program_name) {

	/* show the version. */
	NOTICE("%s (mopaq) %s (zlib %s)\n", program_name, libmpq__version(), zlibVersion());
	NOTICE("Written by %s\n", AUTHOR);
	NOTICE("\n");
	NOTICE("This is free software; see the source for copying conditions.  There is NO\n");
	NOTICE("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");

	/* if no error was found, return zero. */
	return 0;
}

/* this function shows some archive information. */
int mpq_info__archive_info(char *program_name, char *mpq_filename, unsigned int number, unsigned int count) {

	/* some common variables. */
	int result;
	mpq_archive_s *mpq_archive;

	/* allocate memory for the mpq-archive. */
	mpq_archive = malloc(sizeof(mpq_archive_s));
	memset(mpq_archive, 0, sizeof(mpq_archive_s));

	/* open the mpq-archive. */
	if ((result = libmpq__archive_open(mpq_archive, mpq_filename)) < 0) {

		/* open archive failed. */
		NOTICE("archive number:			%i/%i\n", number, count);
		NOTICE("archive name:			%s\n", mpq_filename);
		NOTICE("archive type:			no mpq archive\n");

	} else {

		/* open archive was successful, show information. */
		NOTICE("archive number:			%i/%i\n", number, count);
		NOTICE("archive name:			%s\n", mpq_filename);
		NOTICE("archive version:		%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_VERSION));
		NOTICE("archive size:			%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_SIZE));
		NOTICE("archive hash table entries:	%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_HASHTABLE_ENTRIES));
		NOTICE("archive block table entries:	%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_BLOCKTABLE_ENTRIES));
		NOTICE("archive block size:		%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_BLOCKSIZE));
		NOTICE("archive files:			%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_FILES));
		NOTICE("archive compressed size:	%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_COMPRESSED_SIZE));
		NOTICE("archive uncompressed size:	%i\n", libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_UNCOMPRESSED_SIZE));
		NOTICE("archive compression ratio:	%.2f\n", (100 - ((float)libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_COMPRESSED_SIZE) / (float)libmpq__archive_info(mpq_archive, LIBMPQ_ARCHIVE_UNCOMPRESSED_SIZE) * 100)));
	}

	/* always close file descriptor, file could be opened also if it is no valid mpq archive. */
	libmpq__archive_close(mpq_archive);

	/* free the memory of the mpq-archive. */
	free(mpq_archive);

	/* if multiple archives were given, continue with next one. */
	if (number < count) {
		NOTICE("\n-- next archive --\n\n");
	}

	/* if no error was found, return zero. */
	return 0;
}

/* the main function starts here. */
int main(int argc, char **argv) {

	/* common variables for the command line. */
	int opt;
	int option_index = 0;
	static char const short_options[] = "hv";
	static struct option const long_options[] = {
		{"help",	no_argument,	0,	'h'},
		{"version",	no_argument,	0,	'v'},
		{0,		0,		0,	0}
	};
	optind = 0;
	opterr = 0;

	/* some common variables. */
	char *program_name;
	char mpq_filename[PATH_MAX];
	unsigned int count;
	unsigned int number;

	/* get program name. */
	program_name = argv[0];
	if (program_name && strrchr(program_name, '/')) {
		program_name = strrchr(program_name, '/') + 1;
	}

	/* if no command line option was given, show some info. */
	if (argc <= 1) {

		/* show some info on how to get help. :) */
		ERROR("%s: no action was given\n", program_name);
		ERROR("Try `%s --help' for more information.\n", program_name);

		/* exit with error. */
		exit(1);
	}

	/* parse command line. */
	while ((opt = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1) {

		/* check if all command line options are parsed. */
		if (opt == -1) {
			break;
		}

		/* parse option. */
		switch (opt) {
			case 'h':
				mpq_info__usage(program_name);
				exit(0);
			case 'v':
				mpq_info__version(program_name);
				exit(0);
			default:

				/* show some info on how to get help. :) */
				ERROR("%s: unrecognized option `%s'\n", program_name, argv[optind - 1]);
				ERROR("Try `%s --help' for more information.\n", program_name);

				/* exit with error. */
				exit(1);
		}
	}

	/* fill option structure with long option arguments */
	count = argc - optind;
	number = 1;

	/* create the file count from the command line arguments. */
	do {
		strncpy(mpq_filename, argv[optind], PATH_MAX);
		mpq_info__archive_info(program_name, mpq_filename, number, count);
		number++;
	} while (++optind < argc);

	/* execution was successful. */
	exit(0);
}
